File: lua/nucleotide-count/nucleotide-count.lua
local DNA = {}

function DNA:new(input)
	local dna = {}
	setmetatable(dna, self)
	self.__index = self
	dna.nucleotideCounts = { A = 0, T = 0, C = 0, G = 0 }
	for char in string.gmatch(input, ".") do
		assert(dna.nucleotideCounts[char], "Invalid Sequence")
		dna.nucleotideCounts[char] = dna.nucleotideCounts[char] + 1
	end
	return dna
end

function DNA:count(input)
	return self.nucleotideCounts[input] or error("Invalid Nucleotide")
end

return DNA

File: typescript/nucleotide-count/nucleotide-count.ts
export function nucleotideCounts(input: string): Record<string, number> {
  if (!/^[ACGT]*$/.test(input)) throw new Error('Invalid nucleotide in strand')

  return [...input].reduce(
    (acc: Record<string, number>, char: string) => {
      if (acc[char]) acc[char] += 1
      else acc[char] = 1
      return acc
    },
    {
      A: 0,
      C: 0,
      G: 0,
      T: 0,
    }
  )
}
